fdbclient/SpecialKeySpace.actor.cpp

```diff
@@ -103,47 +103,36 @@ ACTOR Future<Void> moveKeySelectorOverRangeActor(const SpecialKeyRangeBaseImpl*
 	return Void();
 }
 
-void onModuleRead(ReadYourWritesTransaction* ryw, SpecialKeySpace::MODULE module,
-                  Optional<SpecialKeySpace::MODULE>& lastModuleRead) {
-	if (ryw && !ryw->specialKeySpaceRelaxed() && lastModuleRead.present() && lastModuleRead.get() != module) {
-		throw special_keys_cross_module_read();
-	}
-	lastModuleRead = module;
-}
-
 // This function will normalize the given KeySelector to a standard KeySelector:
 // orEqual == false && offset == 1 (Standard form)
 // If the corresponding key is outside the whole space, it will move to the begin or the end
 // It does have overhead here since we query all keys twice in the worst case.
 // However, moving the KeySelector while handling other parameters like limits makes the code much more complex and hard
 // to maintain; Thus, separate each part to make the code easy to understand and more compact
+// Boundary is the range of the legal key space, which, by default is the range of the module
+// And (\xff\xff, \xff\xff\xff\xff) if SPECIAL_KEY_SPACE_RELAXED is turned on
 ACTOR Future<Void> normalizeKeySelectorActor(SpecialKeySpace* sks, ReadYourWritesTransaction* ryw, KeySelector* ks,
-                                             Optional<SpecialKeySpace::MODULE>* lastModuleRead, int* actualOffset,
+                                             KeyRangeRef boundary, int* actualOffset,
                                              Standalone<RangeResultRef>* result,
                                              Optional<Standalone<RangeResultRef>>* cache) {
 	state RangeMap<Key, SpecialKeyRangeBaseImpl*, KeyRangeRef>::Iterator iter =
 	    ks->offset < 1 ? sks->getImpls().rangeContainingKeyBefore(ks->getKey())
 	                   : sks->getImpls().rangeContaining(ks->getKey());
-	while ((ks->offset < 1 && iter != sks->getImpls().ranges().begin()) ||
-	       (ks->offset > 1 && iter != sks->getImpls().ranges().end())) {
-		onModuleRead(ryw, sks->getModules().rangeContaining(iter->begin())->value(), *lastModuleRead);
+	while ((ks->offset < 1 && iter.begin() > boundary.begin) || (ks->offset > 1 && iter.begin() < boundary.end)) {
 		if (iter->value() != nullptr) {
 			wait(moveKeySelectorOverRangeActor(iter->value(), ryw, ks, cache));
 		}
 		ks->offset < 1 ? --iter : ++iter;
 	}
 	*actualOffset = ks->offset;
-	if (iter == sks->getImpls().ranges().begin())
-		ks->setKey(sks->getKeyRange().begin);
-	else if (iter == sks->getImpls().ranges().end())
-		ks->setKey(sks->getKeyRange().end);
+	if (iter.begin() == boundary.begin || iter.begin() == boundary.end) ks->setKey(iter.begin());
```

Comment:
I expected to see `iter->begin()`. I'm surprised that this compiles

Reply:
I think `.` and `->` are both okay here. The iterator object here is defined as
```
class Iterator {
public:
  // ....
  Key const& begin() { return it->key; }
  Iterator* operator->() { return this; }
  //...
private:
  typename Map<Key,Val,pair_type,Metric>::iterator it;
}
```

Reply:
But since it is called `Iterator`, it is against intuition to not use `->`. I will change it.